src/main/java/com/ait/lienzo/client/core/shape/TextTruncateWrapper.java

```diff
@@ -0,0 +1,239 @@
+
+import com.ait.lienzo.client.core.Context2D;
+import com.ait.lienzo.client.core.types.BoundingBox;
+
+/**
+ * ITextWrapper implementation that truncates text and appends "..." if there is no space left.
+ */
+@SuppressWarnings("Duplicates")
+public class TextTruncateWrapper extends TextNoWrap {
+
+    private BoundingBox wrapBoundaries;
+    private final double Y_OFFSET = 0.8;
+
+    public TextTruncateWrapper(final Text text,
+                               final BoundingBox wrapBoundaries) {
+        super(text);
+        setWrapBoundaries(wrapBoundaries);
+    }
+
+    public BoundingBox getWrapBoundaries() {
+        return wrapBoundaries;
+    }
+
+    public TextTruncateWrapper setWrapBoundaries(final BoundingBox boundaries) {
+        wrapBoundaries = boundaries;
+        return this;
+    }
+
+    @Override
+    public BoundingBox getBoundingBox() {
+        final double[] boundaries = calculateWrapBoundaries();
+        return new BoundingBox().addX(0).addX(wrapBoundaries.getWidth()).addY(0).addY(boundaries[1]);
+    }
+
+    private double getWrapBoundariesWidth() {
+        final double margin = 4 * getBoundingBoxForString(" ").getWidth();
+        return this.wrapBoundaries.getWidth() - margin;
+    }
+
+    private double[] calculateWrapBoundaries() {
+        final String[] words = textSupplier.get().split("\\s");
+        if (words.length < 1) {
+            return new double[]{wrapBoundaries.getX(), wrapBoundaries.getY()};
+        }
+
+        final double wrapWidth = getWrapBoundariesWidth();
+        final String firstWord = words[0];
+        double width = getBoundingBoxForString(firstWord).getWidth();
+        final StringBuilder nextLine = new StringBuilder(firstWord);
+        int numOfLines = 1;
+        for (int i = 1; i < words.length; i++) {
+            width = getBoundingBoxForString(nextLine + " " + words[i]).getWidth();
+            if (width <= wrapWidth) {
+                nextLine.append(" ").append(words[i]);
+            } else {
+                nextLine.setLength(words[i].length());
+                nextLine.replace(0,
+                                 words[i].length(),
+                                 words[i]);
+                numOfLines++;
+            }
+        }
+
+        final double lineHeight = (getBoundingBoxForString(" ").getHeight());
+
+        while (!hasVerticalSpace(numOfLines, lineHeight, wrapBoundaries.getHeight() - (Y_OFFSET * numOfLines))
+                && numOfLines >= 0) {
+            numOfLines--;
+        }
+
+        final double height = lineHeight * numOfLines;
+
+        return new double[]{width, height};
+    }
+
+    private boolean hasVerticalSpace(final int lineIndex,
+                                     final double lineHeight,
+                                     final double availableHeight) {
+        return lineHeight * (lineIndex + Y_OFFSET) <= availableHeight;
+    }
+
+    @Override
+    public void drawString(final Context2D context,
```

Comment:
OK, thanks for the explanation... I am wondering (and not tried) but would it be easier to cache the results of the `getBounds()` and `drawString(..)` methods. The cached values could be created when this class is first instantiated and we then add "attribute observers" to recalculate the cached values when certain attributes change; for example:
```
public TextTruncateWrapper(final Text text,
                           final BoundingBox wrapBoundaries) {
    super(text);
    setWrapBoundaries(wrapBoundaries);
    text.addAttributesChangedHandler(Attribute.TEXT,
                                     new AttributesChangedHandler() {
                                         @Override
                                         public void onAttributesChanged(AttributesChangedEvent event) {
                                             recalculate();
                                         }
                                     });
}
```
I'm going to test this PR "as is" (with a graph with lots of nodes on).... :-)

Reply:
Performance seems reasonable enough... happy to keep "as is" - and address any future issues as a separate PR (oh, my comments about reusing the width of a space character - rather than recalculating each time remains something to consider!)